File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetSavedNetworkTests.java
package com.isupatches.wisefy.main;

import static com.isupatches.wisefy.TestUtils.TEST_SSID;
import static com.isupatches.wisefy.TestUtils.VERIFICATION_SUCCESS_TIMEOUT;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;

import android.net.wifi.WifiConfiguration;

import com.isupatches.wisefy.AbstractBaseAndroidJUnit4TestClass;
import com.isupatches.wisefy.WiseFy;
import com.isupatches.wisefy.callbacks.GetSavedNetworkCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodeDefs;

import org.junit.Test;

/**
 * Tests the ability to retrieve a saved on a device.
 *
 * @see WiseFy#getSavedNetwork(String)
 * @see WiseFy#getSavedNetwork(String, GetSavedNetworkCallbacks)
 */
public class GetSavedNetworkTests extends AbstractBaseAndroidJUnit4TestClass {

  public GetSavedNetworkTests() {
    // No-op
  }

  @Test
  public void sync_failure_prechecks() {
    getMockWiseFyPrechecksUtil().getSavedNetwork_failure();
    assertEquals(null, getWiseFy().getSavedNetwork(TEST_SSID));
  }

  @Test
  public void sync_failure() {
    getMockWiseFySearchUtil().findSavedNetworkByRegex_null();
    assertEquals(null, getWiseFy().getSavedNetwork(TEST_SSID));
  }

  @Test
  public void sync_success() {
    final WifiConfiguration savedNetwork = getMockWiseFySearchUtil().findSavedNetworkByRegex_success();
    assertEquals(savedNetwork, getWiseFy().getSavedNetwork(TEST_SSID));
  }

  @Test
  public void async_failure_prechecks() {
    getMockWiseFyPrechecksUtil().getSavedNetwork_failure();
    final GetSavedNetworkCallbacks mockCallbacks = mock(GetSavedNetworkCallbacks.class);
    getWiseFy().getSavedNetwork(TEST_SSID, mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getSavedNetworkWiseFyFailure(WiseFyCodeDefs.MISSING_PREREQUISITE);
  }

  @Test
  public void async_failure_prechecks_nullCallback() {
    getMockWiseFyPrechecksUtil().getSavedNetwork_failure();
    getNullCallbackUtil().callGetSavedNetwork(TEST_SSID);
  }

  @Test
  public void async_failure() {
    getMockWiseFySearchUtil().findSavedNetworkByRegex_null();
    final GetSavedNetworkCallbacks mockCallbacks = mock(GetSavedNetworkCallbacks.class);
    getWiseFy().getSavedNetwork(TEST_SSID, mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).savedNetworkNotFound();
  }

  @Test
  public void async_failure_nullCallback() {
    getMockWiseFySearchUtil().findSavedNetworkByRegex_null();
    getNullCallbackUtil().callGetSavedNetwork(TEST_SSID);
  }

  @Test
  public void async_success() {
    final WifiConfiguration savedNetwork = getMockWiseFySearchUtil().findSavedNetworkByRegex_success();
    final GetSavedNetworkCallbacks mockCallbacks = mock(GetSavedNetworkCallbacks.class);
    getWiseFy().getSavedNetwork(TEST_SSID, mockCallbacks);
    verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetwork(savedNetwork);
  }

  @Test
  public void async_success_nullCallback() {
    getMockWiseFySearchUtil().findSavedNetworkByRegex_success();
    getNullCallbackUtil().callGetSavedNetwork(TEST_SSID);
  }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetSavedNetworkTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.TEST_SSID
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.GetSavedNetworkCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER

import org.junit.Assert.assertEquals
import org.junit.Test

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to retrieve a saved on a device.
 *
 * @author Patches
 */
internal class GetSavedNetworkTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.getSavedNetwork_failure()
        assertEquals(null, wisefy.getSavedNetwork(TEST_SSID))
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        assertEquals(null, wisefy.getSavedNetwork(TEST_SSID))
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        val savedNetwork = mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        assertEquals(savedNetwork, wisefy.getSavedNetwork(TEST_SSID))
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.getSavedNetwork_failure()
        val mockCallbacks = mock(GetSavedNetworkCallbacks::class.java)
        wisefy.getSavedNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetwork_failure()
        nullCallbackUtil.callGetSavedNetwork(TEST_SSID)
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        val mockCallbacks = mock(GetSavedNetworkCallbacks::class.java)
        wisefy.getSavedNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).savedNetworkNotFound()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        nullCallbackUtil.callGetSavedNetwork(TEST_SSID)
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        val savedNetwork = mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        val mockCallbacks = mock(GetSavedNetworkCallbacks::class.java)
        wisefy.getSavedNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedSavedNetwork(savedNetwork)
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.getSavedNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        nullCallbackUtil.callGetSavedNetwork(TEST_SSID)
    }
}
